client-spark/common/src/main/java/org/apache/spark/shuffle/celeborn/SortBasedPusher.java

```diff
@@ -316,11 +379,7 @@ private void allocateMemoryForRecordIfNecessary(int required) throws IOException
 
   @Override
   public long spill(long l, MemoryConsumer memoryConsumer) throws IOException {
-    logger.info("Pushdata in spill, memory used " + getUsed());
-    if (getUsed() != 0) {
-      logger.info("Pushdata is not empty , do push.");
-      return pushData();
-    }
+    logger.warn("SortBasedPusher not support spill yet");
```

Comment:
Curious, why not push data when the `spill` is triggered? @waitinfuture

Reply:
Sorry this reply was pending since May 17...

Reply:
After thoroughly understanding the `pushData()` logic, the `DataPusher`'s buffer is preallocated, which means invoking the `pushData()` method does not result in increased heap memory consumption by the `DataPusher`. I think it's worthy to call `pushData()` to free memory for the  Spark when pipeline is disabled

https://github.com/apache/incubator-celeborn/blob/693172d0bd0614674405f6af863f5f0a361b5f3a/client/src/main/java/org/apache/celeborn/client/write/DataPusher.java#L83-L85

Reply:
Agree

Reply:
When surfacing this part code, I also have similar questions.

> First, when spill is called, we want it truly free memory, but Celeborn's pushData is async, which means even though freeMemory is called inside pushData, the data may still be held in DataPusher before sent to wire. I understand this can be discussed that it's out of Spark's UnifiedMemoryManager.

Async is not a problem when using `CompleteFuture` . Memory free could be called in `spill()` operation

> the thread model has changed. 

From my prospective, just allowing the current memory consumer to trigger spill can solve most problems.

Reply:
> 

Hi @zuston , thanks for your attention on this code!
I reviewed the code again and agree with you and @cfmcgrady that `pushData` can be called in `spill`. It'll be
great if you could raise PRs to improve this (and other places :D)

Reply:
We meet a case of NPE in old code, I think after this pr won't meet this issue. If keep `pushData` here, should add some loc to keep it safe
```
java.lang.NullPointerException
at org.apache.spark.memory.TaskMemoryManager.getPage(TaskMemoryManager.java:407)
at org.apache.spark.shuffle.rss.SortBasedPusher.pushData(SortBasedPusher.java:155)
at org.apache.spark.shuffle.rss.SortBasedPusher.spill(SortBasedPusher.java:317)
at org.apache.spark.memory.TaskMemoryManager.acquireExecutionMemory(TaskMemoryManager.java:177)
at org.apache.spark.memory.TaskMemoryManager.allocatePage(TaskMemoryManager.java:297)
at org.apache.spark.memory.MemoryConsumer.allocatePage(MemoryConsumer.java:116)
at org.apache.spark.sql.execution.python.HybridRowQueue.createNewQueue(RowQueue.scala:228)
at org.apache.spark.sql.execution.python.HybridRowQueue.add(RowQueue.scala:251)
at org.apache.spark.sql.execution.python.EvalPythonExec.$anonfun$doExecute$10(EvalPythonExec.scala:125)
at scala.collection.Iterator$$anon$10.next(Iterator.scala:461)
at scala.collection.Iterator$$anon$10.next(Iterator.scala:461)
at scala.collection.Iterator$GroupedIterator.takeDestructively(Iterator.scala:1161)
at scala.collection.Iterator$GroupedIterator.go(Iterator.scala:1176)
at scala.collection.Iterator$GroupedIterator.fill(Iterator.scala:1214)
at scala.collection.Iterator$GroupedIterator.hasNext(Iterator.scala:1217)

```